### PR TITLE
Move handlers from cmd into separeate package

### PR DESCRIPTION
**Pull request by bliuchak**

### CODE REVIEW

File: internal/platform/storage/couchbase.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+var (
+	ErrUserNotFound = errors.New("user not found")
+)
+
+type Storage struct {
+	cluster *gocb.Cluster
+}
+
+func NewCouchbaseStorage(address, username, password string) (*Storage, error) {
```
Comment (bliuchak) [inline internal/platform/storage/couchbase.go:20]: I'm not really happy with this method.

1) I need somehow to initialize basic database structure it's not existing before. That's why there is weird logic about bucket existence and bucket creation
2) I wasn't expected that there might be timeout in bucket creation.. 😢

File: internal/platform/storage/couchbase.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+	_, err = tokens.Upsert(token, tokenData, 0)
+	if err != nil {
+		return errors.Wrap(err, "can't upset token")
+	}
+
+	return nil
+}
+
+func (s *Storage) DeprecateToken(token Token) error {
```
Comment (bliuchak) [inline internal/platform/storage/couchbase.go:208]: Maybe in this case it's better to just pass the token as string and then just get it from DB and do the stuff?

File: cmd/httpd/handlers/auth.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		a.logger.Error().Err(err).Str("email", user.Email).Msg("Bad password")
+
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	a.logger.Info().Str("email", user.Email).Msg("Successful login")
+
+	expiration := time.Now().Add(30 * time.Minute)
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:79]: Please move 30*time.Minute to const and change to 1 hour

File: cmd/httpd/handlers/auth.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	token, err := a.tokens.CreateToken(user.ID, user.Email, expiration)
+	if err != nil {
+		a.logger.Error().Err(err).Str("email", request.Email).Msg("can't create token")
+
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	a.logger.Info().Str("token", token.Token).Time("exp", expiration).Msg("New token issued")
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:88]: do not put token into log. Potential security issue.

File: cmd/httpd/handlers/auth.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	err := decoder.Decode(&request)
+	if err != nil {
+		a.logger.Error().Err(err).Msg("Unable to decode login data")
+
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// TODO token deprecation probably should be done in transaction
+	oldToken, err := a.tokens.GetNotExpiredTokenByToken(request.Token)
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:117]: We do not store expired tokens. That is why this name is redundant. We can simply use "getTokenByToken"

File: cmd/httpd/handlers/auth.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		if err == storage.ErrTokenNotFound {
+			a.logger.Error().Err(err).Str("token", request.Token).Msg("Old token not exists or already expired")
+
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		a.logger.Error().Err(err).Str("token", request.Token).Msg("Unable to get not expired token")
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:125]: same here: "Unable to get token"

File: cmd/httpd/handlers/auth.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	expiration := time.Now().Add(30 * time.Minute)
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:131]: Can you please refer to the constant for token exp time?

File: cmd/httpd/handlers/auth.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		a.logger.Error().Err(err).
+			Str("user_id", oldToken.Claims.ID.String()).
+			Str("email", oldToken.Claims.Email).
+			Msg("Unable to create new token")
+
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	err = a.tokens.DeprecateToken(oldToken)
```
Comment (grubastik) [inline cmd/httpd/handlers/auth.go:143]: What is the Purpose of this method?

File: internal/platform/storage/couchbase.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	defer users.Close()
+
+	rows, err := users.ExecuteN1qlQuery(query, params)
+	if err != nil {
+		return user, errors.Wrap(err, "can't get user")
+	}
+
+	rows.Close()
```
Comment (grubastik) [inline internal/platform/storage/couchbase.go:174]: probably you forgot defer

File: internal/platform/storage/couchbase.go (hunk 2 above)
Comment (grubastik) [reply internal/platform/storage/couchbase.go:208]: Probaly we need to remove old token.

File: internal/platform/storage/couchbase.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	defer tokens.Close()
+
+	rows, err := tokens.ExecuteN1qlQuery(query, params)
+	if err != nil {
+		return Token{}, errors.Wrap(err, "can't get token")
+	}
+
+	rows.Close()
```
Comment (grubastik) [inline internal/platform/storage/couchbase.go:246]: probably you forgot defer

File: internal/platform/storage/couchbase.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	tokens, err := s.cluster.OpenBucket(tokensBucketName, "")
+	if err != nil {
+		return errors.Wrap(err, "can't open 'tokens' bucket")
+	}
+
+	defer tokens.Close()
+
+	_, err = tokens.Upsert(token, tokenData, 0)
```
Comment (grubastik) [inline internal/platform/storage/couchbase.go:200]: We can use timeout for this record to remove old tokens from DB automatically. If token not exists in DB it is not valid

Comment (grubastik) [reply internal/platform/storage/couchbase.go:200]: Why we even store token in DB?

File: internal/platform/storage/couchbase.go (hunk 2 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:208]: The whole point is whether to store tokens in DB or not.

I assumed that we're storing all tokens in DB. So during token refresh I was needed some method which instantly expire current token. I'm not removing any tokens, just modifying it's expiration time.

File: cmd/httpd/handlers/auth.go (hunk 4 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:88]: Can I than use some token ID instead? Or just don't log any info about token usage?

File: cmd/httpd/handlers/auth.go (hunk 5 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:117]: I'm expiring tokens in this MR.

File: cmd/httpd/handlers/auth.go (hunk 8 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:143]: It's modifying expiration time of token to past date and update DB record.

File: internal/platform/storage/couchbase.go (hunk 9 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:174]: No I don't. I need to execute this method as it is now.

File: internal/platform/storage/couchbase.go (hunk 10 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:246]: No I don't. I need to execute this method as it is now.

File: internal/platform/storage/couchbase.go (hunk 11 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:200]: I was thinking about set expiration on record. But I decided to save tokens and just modify expiration time.

File: cmd/httpd/handlers/auth.go (hunk 4 above)
Comment (grubastik) [reply cmd/httpd/handlers/auth.go:88]: you can mention: token was created for user... but please, do not put token into log

**Comment by grubastik** [comment]
A lot of comments added which are not related to PR. Will create a separate tickets

**Review by grubastik — APPROVED** [review]
New tickets werecreated

File: cmd/httpd/handlers/auth.go (hunk 3 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:79]: ticket for this has been created #6

File: cmd/httpd/handlers/auth.go (hunk 5 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:117]: will be fixed in #5

File: cmd/httpd/handlers/auth.go (hunk 7 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:131]: will be fixed in #6

File: internal/platform/storage/couchbase.go (hunk 9 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:174]: investigation needed #7

File: internal/platform/storage/couchbase.go (hunk 10 above)
Comment (bliuchak) [reply internal/platform/storage/couchbase.go:246]: investigation needed #7

File: cmd/httpd/handlers/auth.go (hunk 4 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:88]: should be fixed in #8

File: cmd/httpd/handlers/auth.go (hunk 6 above)
Comment (bliuchak) [reply cmd/httpd/handlers/auth.go:125]: #8

**Pull request merged**